src/ua_parser/caching.py

```diff
@@ -0,0 +1,121 @@
+import abc
+from collections import OrderedDict
+from typing import Dict, Optional, MutableMapping
+
+from .core import Parser, Domain, PartialParseResult
+
+
+__all__ = [
+    "CachingParser",
+    "Cache",
+    "Clearing",
+    "LRU",
+]
+
+
+class Cache(abc.ABC):
+    """Cache abstract protocol. The :class:`CachingParser` will look
+    values up, merge what was returned (possibly nothing) with what it
+    got from its actual parser, and *re-set the result*.
+
+    A :class:`Cache` is responsible for its own replacement policy.
+    """
+
+    @abc.abstractmethod
+    def __setitem__(self, key: str, value: PartialParseResult):
+        """Adds or replace ``value`` to the cache at key ``key``."""
+        ...
+
+    @abc.abstractmethod
+    def __getitem__(self, key: str) -> Optional[PartialParseResult]:
+        """Returns a partial result for ``key`` if there is any."""
+        ...
+
+
+class Clearing(Cache):
+    """A clearing cache, if the cache is full, just remove all the entries
+    and re-fill from scratch.
+
+    This can also be used as a cheap permanent cache by setting the
+    ``maxsize`` to infinity (or at least some very large value),
+    however this is probably a bad idea as it *will* lead to an
+    ever-growing memory allocation, until every possible user agent
+    string has been seen.
+    """
+
+    def __init__(self, maxsize: int):
+        self.maxsize = maxsize
+        self.cache: Dict[str, PartialParseResult] = {}
+
+    def __getitem__(self, key: str) -> Optional[PartialParseResult]:
+        return self.cache.get(key)
+
+    def __setitem__(self, key: str, value: PartialParseResult):
+        if key not in self.cache and len(self.cache) >= self.maxsize:
+            self.cache.clear()
+
+        self.cache[key] = value
+
+
+class LRU(Cache):
+    """Cache following a least-recently used replacement policy: when
+    there is no more room in the cache, whichever entry was last seen
+    the least recently is removed.
+
+    Note that the cache size is adjusted after inserting the new
+    entry, so the cache will temporarily contain ``maxsize + 1``
+    items.
+    """
+
+    def __init__(self, maxsize: int):
+        self.maxsize = maxsize
+        self.cache: OrderedDict[str, PartialParseResult] = OrderedDict()
+
+    def __getitem__(self, key: str) -> Optional[PartialParseResult]:
+        e = self.cache.get(key)
+        if e:
+            self.cache.move_to_end(key)
+        return e
+
+    def __setitem__(self, key: str, value: PartialParseResult):
+        self.cache[key] = value
+        self.cache.move_to_end(key)
+        while len(self.cache) > self.maxsize:
+            self.cache.popitem(last=False)
+
+
+class CachingParser(Parser):
```

Comment:
Should the caching parser be threadsafe? Or have configurable thread safety?

A pattern of taking a context manager object could work, a thread-safe caching parser would take a Lock while a non-threadsafe cache would use [a `nullcontext`](https://docs.python.org/3/library/contextlib.html?highlight=contextlib#contextlib.nullcontext).

The lock would likely have to be taken on both lookup and insertion: the insertion is when the cache limits get applied, but the lookup might not be thread-safe e.g. on the LRU if the entry we just found gets dropped between line 75 and 77 we get a KeyError. Alternatively, the LRU could just ignore that exception there instead...

Still we might have similar issues with other non-trivial caching policies requiring multiple lookups or somesuch (see https://github.com/ua-parser/uap-python/issues/143#issuecomment-1604631279)